ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+   being searched for.
+b) getFileName() - determines the name of the index being read.
+c) createSegmentMonitors() - defines which columns are being filtered, and what values to match against.
+d) transform() - create the record to return from the activity.  It controls which columns should be included
+   from the index row in the output. (In this case all.)
+
+Executing the graph
+-------------------
+
+To execute a graph, the engine walks the activities in the graph xml and creates in memory a graph of implementation
+activities:
+
+For each activity, the name of the helper factory is calculated from the activity number (e.g., fAc2 for
+activity 2), that function is imported from the loaded dll, and then called to create an instance of the generated helper
+class - in this case cAc2.
+
+The engine creates an instance of the class for implementing that kind of activity.  (E.g., activity 2 
+has a _kind of 77 - which corresponds to TAKindexread, so roxie will create an instance of CRoxieServerIndexReadActivity)
+and pass the previously created helper object (which must implement IHThorIndexReadArg) to the constructor.  Some
+activity implementations may also extract other information from the xml for the activity - e.g., hints.  Once all the
+activities are created the edge information is used to link inputs to outputs and add other dependencies.  
+
+Note: Any subgraph that is marked with <att name="rootGraph" value="1"/> is a root subgraph.  An activity within
+a subgraph that has no outputs is called a sink (activities without inputs are called sources).
+
+Executing a graph involves executing all the root subgraphs that it contains.  All
+dependencies of the activities within the subgraph must be executed before a subgraph is executed.  
+To execute a subgraph, the engine executes each of the sink activities
+on separate threads, and then waits for each of the threads to complete.  
+Each sink activity lazily pulls input rows on demand from activities further up the graph, processes than and returns
+when complete.
+
+(If you examine the code you will find the implementation for processing dependencies is more fine grained, to ensure
+IF datasets, OUPUT(,UPDATE) and other ECL constructs are executed correctly.)
+
+In our example the execution will flows as follows:
+
+1. Only a single root subgraph, so need to execute that.
+2. The engine will execute activity 3 the workunit write activity (TAKworkunitwrite).
+3. The workunit write activity will start its input.
+4. The index read activity will call the helper functions to get the filename, resolve the index, and create the filter.
+5. The workunit activity requests a row from its input.
+6. The index read finds the first row, and calls the helper's transform() method to create an output row.
+7. The workunit activity saves that row away (possibly using the serializer provided by the IOutputMetaData interface
+   implemented by the class mx1.
```

Comment:
I think it might be a good idea that proper nouns are encapsulated in single quotations throughout this document.